### DISCUSSION TITLE
Beatmaps not being read

### DISCUSSION THREAD

**Discussion by aaryanajaysharma** [ppy/osu#24716] [Q&A]
After running 2021 version of osu!lazer on my Macbook Air, when I reran the latest version, all my beatmaps were gone. I do not know how this happened, but it seems like the osu!lazer client is not reading the beatmaps that are present in the osu directory.

It seems so because the files folder seem to be intact. Here's a screenshot of my osu directory:
<img width="1162" alt="image" src="https://github.com/ppy/osu/assets/78302134/e4b7c0ab-9288-4792-a540-367fe4080a66">

> logs

[database.log](https://github.com/ppy/osu/files/12519275/database.log)
[input.log](https://github.com/ppy/osu/files/12519276/input.log)
[network.log](https://github.com/ppy/osu/files/12519277/network.log)
[performance.log](https://github.com/ppy/osu/files/12519278/performance.log)
[runtime.log](https://github.com/ppy/osu/files/12519279/runtime.log)

> client files

[client.zip](https://github.com/ppy/osu/files/12519359/client.zip)

**Comment by peppy** [comment]
Hopefully you have a backup of your `client.db`. osu! may have created one for you, but you'll have to check the files in that folder to see if that's the case. The ones you included in your post are definitely nuked.

I wouldn't really recommend running old versions out of the blue like this... we had a lot less safeties in the game back then.

**Reply by aaryanajaysharma** [reply]
Thanks a lot, I'll try not running the older version. Would the backup be located somewhere in the directory?

And let's say there isn't a backup of the client.db, would it be possible to recover the files then? Since all the beatmaps are reliably  stored and present in the files folder,

**Reply by peppy** [reply]
If you don't have a copy of your old database, it won't be possible to recover. The files folder does include files, but all the metadata related to beatmaps is in the main database.

**Reply by aaryanajaysharma** [reply]
Thanks. I'm not sure what the name of the backup db might be. Is it `online.db` or `collection.db`? If I knew that, I'd be able to do something.

**Reply by peppy** [reply]
client.something.realm probably

**Reply by aaryanajaysharma** [reply]
I fooled around a little and got to know the backup file to `client.realm` was `client.v22.backup.realm` which was overwritten as soon as opened the 2021 version of osu!lazer. 

- I think this is a problem. Opening some of the previous version(s) of the game overwrites the client database (specifically `client.realm` file and `client.v22.backup.realm`) file, which should not be the case. Even though all the files are securely stored, if the client database which contains mappings to the data, is lost, it effectively results in the loss of all the users' data. 
   - Should I turn this discussion to an issue, if this is something that needs to be fixed?
- Now that it's confirmed that the data (beatmaps, skins, replays, scores) has effectively been lost, is there a way to recover them? I'd be very grateful if you could provide an alternative/workaround to this since I put a considerable amount of time and effort to bring the game to this stage.
- The lazer client as been glitching a lot while I'm playing, and in general not working as expected since I've messed around with system files of the game. Is there something I can do to fix it?

Thanks a lot for answering.

**Reply by peppy** [reply]
Not sure sorry. If you provide the whole folder zipped (minus the `files` directory) I can check one last time, but I think you're out of luck.

**Reply by aaryanajaysharma** [reply]
Sure. Here's the link to the zip: https://drive.google.com/file/d/1-NFO24MqbOWRPjy9lhQFWRSf2muqtwJE/view?usp=sharing

Thanks again! If nothing works, I'll just try reinstalling the game altogether.

**Reply by peppy** [reply]
You still have data from your older database present, so you may be able to recover  a good portion.

- Download this release: https://github.com/ppy/osu/releases/tag/2022.911.0
- Delete all `.realm` files from the folder. **Do not delete `client.db`**
- Let the game migrate your data from the `client.db` to a `client.realm` file

Once this is done, you should be able to run the latest release.

**Reply by aaryanajaysharma** [reply]
The data is still not showing up in the client.

After downloading the release and deleting the realm files in the directory, when I try to open the osu!lazer client, another folder called `osu-lazer` gets created within the directory.

Here's the image of the directory before opening the client:
<img width="761" alt="image" src="https://github.com/ppy/osu/assets/78302134/eddd30b4-3569-4eae-a036-793801fa0c8d">

And here it is after opening the client:
<img width="760" alt="image" src="https://github.com/ppy/osu/assets/78302134/8ebe3e77-7536-47be-a4b8-77c1ba62bdbc">

logs of the parent directory which has the files folder are unchanged.

Here are the logs within the `osu-lazer` directory:
[database.log](https://github.com/ppy/osu/files/12555741/database.log)
[network.log](https://github.com/ppy/osu/files/12555743/network.log)
[performance.log](https://github.com/ppy/osu/files/12555744/performance.log)
[runtime.log](https://github.com/ppy/osu/files/12555745/runtime.log)

Here is the output for the `tree -I 'files*|cache*'` command:

```.
├── client.db
├── client.realm.note
├── client.zip
├── collection.db
├── exports
│   ├── Guest playing Fujiwara Chika (CV_ Kohara Konomi) - Chikatto Chika Chika (hypercyte) [Normal] (2023-07-29_04-47).osr
│   └── aaryansharma playing Ai Higuchi - Akuma no Ko (TV Size) (Aakki) [Easy] (2023-07-24_10-15).osr
├── framework.ini
├── game.ini
├── input.json
├── logs
│   ├── database.log
│   ├── input.log
│   ├── network.log
│   ├── performance.log
│   └── runtime.log
├── online.db
├── osu-lazer
│   ├── client.realm
│   ├── client.realm.lock
│   ├── client.realm.management
│   │   ├── access_control.control.mx
│   │   ├── access_control.new_commit.cv
│   │   ├── access_control.pick_writer.cv
│   │   └── access_control.write.mx
│   ├── client.realm.note
│   ├── framework.ini
│   ├── game.ini
│   ├── input.json
│   ├── logs
│   │   ├── database.log
│   │   ├── network.log
│   │   ├── performance.log
│   │   └── runtime.log
│   ├── online.db
│   ├── rulesets
│   └── screenshots
├── osu.tgz
├── rulesets
├── screenshots
├── storage.ini
└── zip
    ├── client.db
    ├── client.realm
    ├── client.realm.backup-log
    ├── client.realm.lock
    ├── client.realm.management
    │   ├── access_control.control.mx
    │   ├── access_control.versions.mx
    │   └── access_control.write.mx
    ├── client.v22.backup.realm
    ├── collection.db
    ├── exports
    │   ├── Guest playing Fujiwara Chika (CV_ Kohara Konomi) - Chikatto Chika Chika (hypercyte) [Normal] (2023-07-29_04-47).osr
    │   └── aaryansharma playing Ai Higuchi - Akuma no Ko (TV Size) (Aakki) [Easy] (2023-07-24_10-15).osr
    ├── framework.ini
    ├── game.ini
    ├── input.json
    ├── logs
    │   ├── database.log
    │   ├── input.log
    │   ├── network.log
    │   ├── performance.log
    │   └── runtime.log
    ├── online.db
    ├── osu.zip
    ├── rulesets
    ├── screenshots
    └── storage.ini

16 directories, 54 files
```

I started playing just over a month ago, and there were only about 250 beatmaps that I downloaded, so I just brute forced all the beatmap downloads by going through my play history. If this doesn't work, I'd be alright since I already have majority of my beatmaps with me now. It'd be nice to have to have all those replays back too but alas.

If there is anything that can be done next, please let me know. And thanks a lot for the help!

**Reply by peppy** [reply]
Try putting your client.db inside the `osu-lazer` folder. But to reiterate, this will recover only your data from <2022 and not the more recent stuff, so if you care more about the recent stuff it's probably better to start fresh.

One thing to note is that your `files` directory still has all the raw file data for the beatmaps, so it will be taking up quite a lot more disk space than required. When you redownload beatmaps it won't make a duplicate of files, so it's not as bad as it could be. We'll eventually provide a way to safely clean up the files folder in the future.
